### ISSUE TITLE
Alternate Encryption Architecture and Secret Sharing Protocol

### ISSUE THREAD

**Issue by MattSurabian** [bocoup/secrets#21]
The current crypto setup in #19 is fine for local command line use and is cryptographically secure. However it doesn't make automated sharing very straightforward when multiple users are involved. The below suggestion addresses that problem and prevents the need for out of band password sharing. 

This makes it much easier to automate secret sharing to authorized clients.
 
**General Prerequisites:**
 - Each user has a TLS client cert and key for authentication
 - Each user has an RSA public/private key pair
 - Each user has pushed an anonymized version of their public key (comments stripped) to the server.
  
**Assumed Prerequisites For This Example:**
  - A vault `V` has already been created and Users A, B, C have access to it.

**When User A  runs the `create` command in their client to push a new secret `s1` to the vault `V` the following occurs:**
  1. The client asks the server for all the public keys of the users which currently have access to `V`. The server responds with `kA`, `kB`, `kC`.
  1. The client generates a new public/private keypair `pubk1` and `privk1` which will be used to encrypt the secret `s1`
  1. The client encrypts the `s1` with `pubk1` creating the cipherText `s1ct`
  1. The client determines the length `L` of `s1ct`
  1. The client encrypts `privk1` with `kA`, `kB`, and `kC` creating `pk1A`, `pk1B`, `pk1C`
  1. The client generates fingerprints `fA`, `fB`, `fC` for `kA`, `kB`, `kC` with `sha256` 
  1. The client securely destroys `privk1`
  1. The encrypted payload sent to the server is assembled by concatenating: `L s1ct fA pk1A fB pk1B fC pk1C`

This construction allows all authorized users to easily decrypt the secret at a later time while ensuring that the server never receives any of the private key data. In the event a new user is added to the vault an update command could easily be run by any existing authorized user.

**Comment by MattSurabian** [comment]
Thoughts on this approach @mzgoddard @tafsiri @gnarf @calvinmetcalf?

**Comment by calvinmetcalf** [comment]
> The client generates a new public/private keypair pubk1 and privk1 which will be used to encrypt the secret s1

this should just be that the client generates a random secret and maybe signs it with their own private key.

no reason to use public key encryption when you are just encrypting the key, you can use secret key encryption for that.

**Comment by MattSurabian** [comment]
@calvinmetcalf That's true, the actual secret encryption can stay GCM with a randomly generated key.

**Comment by tkellen** [comment]
@tarcieri would you have any time to weigh in on this? @MattSurabian do we have a link that succinctly explains what this project aims to do?

**Comment by tarcieri** [comment]
Some prior art, FWIW:

https://corner.squareup.com/2013/10/parliament-a-secret-sharing-service.html
https://github.com/livingsocial/keyspace

**Comment by MattSurabian** [comment]
@tkellen I thought it was in the docs folder but it is not, I'll be sure to add the high level project description. 

Thanks for the links @tarcieri, I looked over them both briefly and of the two keyspace is the closest to what we're trying to accomplish. Parliament, while interesting, requires that secrets on the server can be "unlocked" at certain times or that users push their private key to allow decryption which is something I principally disagree with.

Essentially the main use case for this project is asynchronous secret sharing between multiple parties. More specifically server deployment. Using built in configuration management tool options (chef encrypted data bags, nordstrom's vault, ansible encrypted files etc) is frustrating and requires secret duplication in large environments with diverse tooling. Likewise having to bake AMIs or basebox's with secrets on them is insecure. The goal here is to be able to create a base box with as little sensitive information on it as possible and scale out multiple servers while not sacrificing the ability to revoke access to specific clients at any given time.
